### ISSUE TITLE
Raycasting attached entities may return incorrect objectref or fail entirely

### ISSUE THREAD

**Issue by GreenXenith** [minetest/minetest#10304]
(Version 5.3.0 stable)

When raycasting to an attached entity, somtimes the raycast just doesn't find anything, or the object reference it returns is that of an entirely different entity (presumably that of another child entity or maybe the parent .. I haven't experienced consistent behavior yet). All other data (intersection point, normal, etc) is correct when something is actually returned.

**Comment by hecktest** [comment]
Does a pure dedicated server still link irrlicht? If yes, just make it create a mock scene to do those things with.
If not, well, this means we need our own animator and hierarchy solver. And scene importers, at least the transform part. This _is_ doable, it's just grunt work. It would also be a major step towards ditching irrlicht entirely.

**Comment by SmallJoker** [comment]
Server builds only require the Irrlicht headers for classes like `vector3d` or `u32`. Making Irrlciht a hard dependency for server builds would solve that. I don't think that would be a too big deal if `libirrlicht-dev` is installed anyway.

**Comment by rubenwardy** [comment]
I think we should be reducing dependency on Irrlicht, not increasing

**Comment by hecktest** [comment]
Right, irrlicht itself depends on X and that causes trouble when trying to run a true headless server.
Native animation and transforms it is... feel free to ninja me on it, I have no plans for this at the moment. But if this is ever done, it would be logical to keep chipping away at irrlicht dependency until none is left. (it's already mostly being used as a crappy OpenGL shim, take away scenes and animation and there really isn't much left)

**Comment by Zughy** [comment]
Up for visibility. This is a really annoying issue that prevents modders from doing many things with entities. E.g. having a barrier attached to the player that blocks raycasts (lasers, bullets and so on)

**Comment by SmallJoker** [comment]
Unless bone positions can be read in by the server (they yet cannot), they would have to be specified manually by using an approximation. In addition to that, the collision/raycast code does currently not handle rotated cuboids, hence actual shield functionality like what you seem to describe will be rather difficult to implement properly.

**Comment by fluxionary** [comment]
> Unless bone positions can be read in by the server (they yet cannot), they would have to be specified manually by using an approximation. 

SSCSM?

**Comment by GreenXenith** [comment]
I've updated this issue as part of the problem (collisions) was solved at some point. Raycasts still need attention. I have opened issue #14696 that addresses the root problem with other feature considerations. This issue would likely be solved when that one is.
